### PR TITLE
Add Associations classes and Methods

### PR DESCRIPTION
**Pull request by Cosmosha**
### The following requirement or features were added;

- Gitflow
- Linters
- Add Classroom Class
- Add Rentals Class
- Add Book Class
- Create has-many / belong-to relationsip between Classes

### CODE REVIEW

File: classroom.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class Classroom
+  attr_accessor :label
+
```
Comment (Shedrack-Sunday) [inline classroom.rb]: - [ ] In implementing  `has-many / belongs-to` relationships between classroom and Student, kindly add an attribute reader to access `students`

File: student.rb [hunk 2]
```diff
@@ -9,4 +9,9 @@ def initialize(age, classroom, name = 'Unknown', parent_permission: true)
   def play_hooky
     '¯\(ツ)/¯'
   end
+
+  def add_classroom(classroom)
+    @classroom = classroom
+    classroom.students << self unless classroom.students.include?(self)
+  end
```
Comment (Shedrack-Sunday) [inline student.rb:18]: - [ ] When setting the classroom for a student it also adds it to the classrooms' students. You're missing an accessor for `classroom`

File: book.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals << rental
+  end
```
Comment (Shedrack-Sunday) [inline book.rb]: - [ ] The book itself should be sent as a parameter to create the new `rental` achieving the 'has-many' association.
- [ ] Kindly review the Link to the example in a previous lesson: https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship`

### Sample
```
def add_rental(person, date)
    Rental.new(date, self, person)
```

File: rental.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class Rental
+  attr_accessor :date, :student, :book
+
+  def initialize(date)
+    @date = date
+    @person = person
+    person.add_rental(self)
+    @book = book
+    book.add_rental(self)
```
Comment (Shedrack-Sunday) [inline rental.rb]: - [ ] In achieving "A rental belongs to a book and also belongs to a person", you need to fix a  couple of changes here:
- The constructor of `Rental` should be modified so `Book and Person` are set in it to break the `many-to-many `relationship. Link to the example article: https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship
  
  ### Sample code:
```
def initialize(date, book, person)

    @book = book
    # The rental is adding itself to the @book achieving the 'belongs-to' association
    book.rentals << self

    @person = person
    # The rental is adding itself to the @person the 'belongs-to' association
    person.rentals << self
  end
```

File: rental.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+class Rental
+  attr_accessor :date, :student, :book
```
Comment (Shedrack-Sunday) [inline rental.rb]: - [ ] For this section,  Getters and Setters for `@person` instance variables should be present. Kindly refer to this  [resource](https://github.com/microverseinc/curriculumruby/blob/main/oop/articles/oop_principles_examples.md#useful-shortcuts-attr_reader-attr_writer-and-attr_accessor)

File: person.rb [hunk 6]
```diff
@@ -21,6 +20,10 @@ def correct_name
     @name
   end
 
+  def add_rental(rental)
+    @rentals << rental
+  end
+
```
Comment (Shedrack-Sunday) [inline person.rb]: - [ ] Firstly, kindly add  getters and setters for `rentals`  as a `person` should have many rentals 
- [ ] Secondly, `@rentals`  should be initialized  in your constructor Link to the example about constructors in a previous [lesson]( https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/class_and_object_for_irb.md#constructor-and-instance-vars)
- [ ] Thirdly, The `person` itself should be sent as a parameter to create the new rental achieving the 'has-many' association. For that, add a `date` to the parameter of the method. This should refactor your method to somthing like this:
### Sample
```
Rental.new(date, book, self)
end
```

**Review by Shedrack-Sunday — CHANGES_REQUESTED** [review]
Hello @Cosmosha 👋 

# CHANGES REQUIRED ♻️

Great work on the project 👏
You've done extremely well so far in this project,  However, there are some adjustments you need to make to move to the next project. 💯



## Highlights   
- Good use of GitHub features ✔️  
- Nice attempt in trying to meet the requirements ✔️ 
  
# Required Changes  🌵 

- [ ] You made some impressive implementations which I must say is a job well done. However, there are a few  I suggest   you make In line with the requirement


_Check the comments for a detailed review 

# Optional suggestions 

Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing really improves your application.

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.

As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: rental.rb (hunk 5 above)
Comment (Cosmosha) [reply rental.rb]: The link to the resource is not working, can you please reshare the link.

File: classroom.rb (hunk 1 above)
Comment (rloterh) [reply classroom.rb]: well resolved! ✔️

File: student.rb (hunk 2 above)
Comment (rloterh) [reply student.rb:18]: well resolved! ✔️

File: book.rb (hunk 3 above)
Comment (rloterh) [inline book.rb:12]: - [ ] Nice job so far creating the `add_rental` method, however, the _book_ itself should be sent as a parameter to create the new _rental_, thereby achieving the _'has-many'_ association. We can achieve this by modifying our code to look like the following example below:

```ruby
class Book
  attr_accessor :title, :author, :rentals
  .
  .
  def add_rental(person, date)
    Rental.new(date, self, person)
  end
end
```

Comment (rloterh) [reply book.rb]: Needs attention!

File: rental.rb (hunk 4 above)
Comment (rloterh) [reply rental.rb]: well resolved! ✔️

File: rental.rb (hunk 5 above)
Comment (rloterh) [reply rental.rb]: well resolved! ✔️

File: person.rb (hunk 6 above)
Comment (rloterh) [inline person.rb:26]: - [ ] Here again, let's ensure the _person_ itself is sent as a parameter to create the new _rental_, thereby achieving the _'has-many'_ association. We can achieve this by modifying our code to look like the following example below:

```ruby
class Person
  attr_accessor :name, :age, :rentals
  .
  .
  def initialize(age, name = "Unknown", parent_permission = true)
    .
    .
   @rentals = []
  end
  .
  .

  def add_rental(book, date)
    Rental.new(date, book, self)
  end

  .
  .
end
 ```

- [ ] In addition, let's mark `rental` with `attr_acessor` and initialize @rentals in a constructor, like the code above.

Comment (rloterh) [reply person.rb]: Needs attention!

File: rental.rb (hunk 5 above)
Comment (rloterh) [reply rental.rb]: Resource link [here](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_principles_examples.md#useful-shortcuts-attr_reader-attr_writer-and-attr_accessor)

File: person.rb (hunk 6 above)
Comment (rloterh) [inline person.rb]: - [ ] Here again, let's ensure the _person_ itself is sent as a parameter to create the new _rental_, thereby achieving the _'has-many'_ association. We can achieve this by modifying our code to look like the following example below:

```ruby
class Person
  attr_accessor :name, :age, :rentals
  .
  .
  def initialize(age, name = "Unknown", parent_permission = true)
    .
    .
   @rentals = []
  end
  .
  .

  def add_rental(book, date)
    Rental.new(date, book, self)
  end

  .
  .
end
 ```

- [ ] In addition, let's mark `rental` with `attr_acessor` and initialize @rentals in a constructor, like the code above.

File: book.rb (hunk 3 above)
Comment (rloterh) [inline book.rb]: - [ ] Nice job so far creating the `add_rental` method, however, the _book_ itself should be sent as a parameter to create the new _rental_, thereby achieving the _'has-many'_ association. We can achieve this by modifying our code to look like the following example below:

```ruby
class Book
  attr_accessor :title, :author, :rentals
  .
  .
  def add_rental(person, date)
    Rental.new(date, self, person)
  end
end
```

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Cosmos, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there!
🏁 🚦 🏃 🌴 🏫 🌳 🏘️

## Highlights 👍  
- Good job, object associations are taking shape!
- ReadMe looks!

## Required Changes ♻️
_Kindly check the inline comments under the review._

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by DeliceLydia — APPROVED** [review]
## Hi @Cosmosha, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
# Congratulations! 🎉 

## Highlights
- Addressed all changed requestef


### Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**